### PR TITLE
Pr checkpoint

### PR DESCRIPTION
**Pull request by Haichao-Zhang**

### CODE REVIEW

File: alf/utils/checkpoint_utils.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            checkpoint = torch.load(f_path)
+            _load_checkpoint(checkpoint)
+        else:
+            warnings.warn(("Checkpoint 'ckpt-{}' does not exist. "
+                           "Train from scratch.".format(global_step)))
+
+        return self._global_step + 1
+
+    def save(self, global_step):
+        """Save states of all modules to checkpoint
```
Comment (emailweixu) [inline alf/utils/checkpoint_utils.py]: comment global_step

File: alf/utils/checkpoint_utils.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        self._modules = kwargs
+        self._ckpt_dir = ckpt_dir
+        self._global_step = -1
+
+        os.makedirs(self._ckpt_dir, exist_ok=True)
+
+    def load(self, global_step="latest"):
+        """Load checkpoint
+        Args:
+            global_step: the number of training steps corresponding to the
```
Comment (emailweixu) [inline alf/utils/checkpoint_utils.py]: comment global_step="latest"
global_step (type)

File: alf/utils/checkpoint_utils.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            checkpoint = torch.load(f_path_latest)
+            _load_checkpoint(checkpoint)
+        elif os.path.isfile(f_path):
+            checkpoint = torch.load(f_path)
+            _load_checkpoint(checkpoint)
+        else:
+            warnings.warn(("Checkpoint 'ckpt-{}' does not exist. "
+                           "Train from scratch.".format(global_step)))
+
+        return self._global_step + 1
```
Comment (emailweixu) [inline alf/utils/checkpoint_utils.py]: should be self._global_step

Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: Previously designed to return the ```next_train_step_num```. Now change to return the ```curretn_train_step_num``` according to suggestion.

File: alf/utils/checkpoint_utils.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: Comments added.

File: alf/utils/checkpoint_utils.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: Comments added.

File: alf/utils/checkpoint_utils.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        Returns:
+            next_step_num (int): the step_number for the next training step
+        """
+
+        def _load_checkpoint(checkpoint):
+            self._global_step = checkpoint["global_step"]
+            for k, v in self._modules.items():
+                self._modules[k].load_state_dict(checkpoint[k])
+
+        f_path_latest = os.path.join(self._ckpt_dir, "latest")
```
Comment (emailweixu) [inline alf/utils/checkpoint_utils.py]: handle the case if latest does not exist.

File: alf/utils/checkpoint_utils.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        """
+        self._global_step = global_step
+        f_path = os.path.join(self._ckpt_dir, "ckpt-{0}".format(global_step))
+        state = {
+            k: v.module.state_dict()
+            if type(v) == torch.nn.DataParallel else v.state_dict()
+            for k, v in self._modules.items()
+        }
+        state['global_step'] = global_step
+        torch.save(state, f_path)
```
Comment (emailweixu) [inline alf/utils/checkpoint_utils.py]: write a message for successful loading

Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: now added message for both successful loading and saving

File: alf/utils/checkpoint_utils.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: Line 55 checks for the existence of latest, and if it does not exist, a warning will be given (Line 62).

**Comment by emailweixu** [comment]
Can you make sure we can checkpoint Algorithm in all kinds of scenarios?

**Comment by Haichao-Zhang** [comment]
Improved the code and added test cases for various potential use cases, including sub-algorithm with optimizer, ignored sub-algorithm, cycle, parameter-sharing etc.

File: alf/algorithms/algorithm.py [hunk 6]
```diff
@@ -310,6 +322,101 @@ def convert_train_state_to_predict_state(self, state):
...
+                and optimizers' state (if any)
+
+        Args:
+            destination (OrderedDict): the destination for storing the state
+        Returns:
+            destination (OrderedDict): the dictionary including both model state
+                and optimizers' state (if any)
+
+        """
+        self._assert_no_cycle_or_duplicate(ignore=False)
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: with ignore=False, the assertion may fail.

File: alf/algorithms/algorithm.py [hunk 7]
```diff
@@ -309,6 +314,242 @@ def convert_train_state_to_predict_state(self, state):
...
+                            visited=None):
+        r"""Saves module state to `destination` dictionary, containing a state
+        of the module, but not its descendants. This is called on every
+        submodule in :meth:`~torch.nn.Module.state_dict`.
+        In rare cases, subclasses can achieve class-specific behavior by
+        overriding this method with custom logic.
+        Arguments:
+            destination (dict): a dict where state will be stored
+            prefix (str): the prefix for parameters and buffers used in this
+                module
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: comment keep_vars

File: alf/algorithms/algorithm.py [hunk 8]
```diff
@@ -309,6 +314,242 @@ def convert_train_state_to_predict_state(self, state):
...
+                    load(child, prefix + name + '.', visited=visited)
+
+            local_metadata = {} if metadata is None else metadata.get(
+                prefix[:-1], {})
+            module._load_from_state_dict(state_dict, prefix, local_metadata,
+                                         True, missing_keys, unexpected_keys,
+                                         error_msgs, visited)
+
+        load(self)
+        load = None  # break load->load reference cycle
```
Comment (emailweixu) [inline alf/algorithms/algorithm.py]: Don't understand why this is needed.

File: alf/algorithms/algorithm.py [hunk 9]
```diff
@@ -309,6 +314,242 @@ def convert_train_state_to_predict_state(self, state):
...
+            3) optimizers: save the optimizers for all the (sub-)algorithms
+
+        Args:
+            destination (OrderedDict): the destination for storing the state
+        Returns:
+            destination (OrderedDict): the dictionary including both model state
+                and optimizers' state (if any)
+
+        """
+        #self._assert_no_cycle_or_duplicate()
```
[the remark this answers is not in the document]
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: removed

File: alf/algorithms/algorithm.py [hunk 10]
```diff
@@ -309,6 +314,242 @@ def convert_train_state_to_predict_state(self, state):
...
+                   keep_vars=False,
+                   visited=None):
+        """Get state dictionary recursively, including both model state
+        and optimizers' state (if any). It can handle a number of special cases:
+            1) graph with cycle: save all the states and avoid infinit loop
+            2) parameter sharing: save only one copy of the shared module/param
+            3) optimizers: save the optimizers for all the (sub-)algorithms
+
+        Args:
+            destination (OrderedDict): the destination for storing the state
```
[the remark this answers is not in the document]
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: comment added

File: alf/algorithms/algorithm.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: comment added

File: alf/algorithms/algorithm.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: It's unnecessary. Removed

File: .ci-cd/build.sh [hunk 11]
```diff
@@ -49,13 +49,15 @@ function test() {
...
+        alf.networks.value_networks_test \
         alf.networks.encoding_networks_test \
         alf.networks.projection_networks_test \
         alf.tensor_specs_test \
+        alf.utils.checkpoint_utils_test \
         alf.utils.dist_utils_test \
         alf.utils.value_ops_test \
-        alf.networks.actor_distribution_networks_test \
-        alf.networks.value_networks_test
+
```
Comment (Haichao-Zhang) [inline .ci-cd/build.sh]: Tests are reordered

File: alf/algorithms/algorithm.py [hunk 12]
```diff
@@ -309,6 +314,247 @@ def convert_train_state_to_predict_state(self, state):
...
 
+    @common.add_method(nn.Module)
+    def state_dict(self,
+                   destination=None,
+                   prefix='',
+                   keep_vars=False,
+                   visited=None):
+        """Get state dictionary recursively, including both model state
+        and optimizers' state (if any). It can handle a number of special cases:
+            1) graph with cycle: save all the states and avoid infinit loop
```
Comment (hnyu) [inline alf/algorithms/algorithm.py]: infinit -> infinite

File: alf/algorithms/algorithm.py [hunk 13]
```diff
@@ -309,6 +314,247 @@ def convert_train_state_to_predict_state(self, state):
...
+            2) parameter sharing: save only one copy of the shared module/param
+            3) optimizers: save the optimizers for all the (sub-)algorithms
+
+        Args:
+            destination (OrderedDict): the destination for storing the state
+            prefix (str): a string to be added before the name of the items
+                (modules, params, algorithms etc) as the key used in the
+                state dictionary
+            keep_vars (bool): detach from params and buffers if False
+            visited (set): a set keeping track of the visited objects
```
Comment (hnyu) [inline alf/algorithms/algorithm.py]: This `visited` is basically an internal argument that's never used by a normal user. A better way is to remove this argument and define another child function inside for the recursion.

File: alf/algorithms/algorithm.py [hunk 14]
```diff
@@ -309,6 +314,247 @@ def convert_train_state_to_predict_state(self, state):
...
+        return destination
+
+    def load_state_dict(self, state_dict, strict=True):
+        """Load state dictionary for Algorithm
+        Arguments:
+            state_dict (dict): a dict containing parameters and
+                persistent buffers.
+            strict (bool, optional): whether to strictly enforce that the keys
+                in :attr:`state_dict` match the keys returned by this module's
+                :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
```
Comment (hnyu) [inline alf/algorithms/algorithm.py]: comment on what happens if `strict=False`

File: alf/utils/checkpoint_utils.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        """A class for making checkpoints.
+
+        Args:
+            ckpt_dir: The directory to save checkpoints. Create ckpt_dir if
+                it doesn't exist.
+            kwargs: Items to be included in the checkpoint. Each item needs
+                to have state_dict and load_state_dict implemented.
+                For instance of Algorithm, only the root need to be passed in,
+                all the children modules and optimizers are automatically
+                extracted and checkpointed.
```
Comment (hnyu) [inline alf/utils/checkpoint_utils.py]: Comment on what happens if children modules are also passed in

File: alf/utils/checkpoint_utils_test.py [hunk 16]
```diff
@@ -0,0 +1,549 @@
...
+        self.assertTrue((alg_root2.state_dict() == expected_state_dict))
+
+
+class TestModelMismatch(unittest.TestCase):
+    def test_model_mismatch(self):
+        # test model mis-match
+
+        ckpt_dir = "/tmp/ckpt_data/model_mis_match/"
+
+        if os.path.isdir(ckpt_dir):
```
Comment (hnyu) [inline alf/utils/checkpoint_utils_test.py]: For all the tests that use tmp files, suggest using `tempfile` context manager that will automatically create and delete tmp directories.

File: alf/algorithms/algorithm.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: corrected

File: alf/algorithms/algorithm.py (hunk 14 above)
Comment (Haichao-Zhang) [reply alf/algorithms/algorithm.py]: comments added

File: alf/utils/checkpoint_utils.py (hunk 15 above)
Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils.py]: added

File: alf/utils/checkpoint_utils_test.py (hunk 16 above)
Comment (Haichao-Zhang) [reply alf/utils/checkpoint_utils_test.py]: good suggestion! changed

**Pull request merged**